cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java

```diff
@@ -974,17 +995,72 @@ public void removeNullOrEmptyTags(final MetadataEntity metadataEntity) {
       Strings.isNullOrEmpty(getMetadata(metadataEntity, input).getValue()));
   }
 
+  /**
+   * Add new key-value pair as metadata
+   *
+   * @param metadataEntry The value of the metadata to be saved
+   * @return {@link MetadataChange} representing the metadata before the change and after
+   */
+  private MetadataChange setMetadata(MetadataEntry metadataEntry) {
+    Set<Indexer> indexersForKey = getIndexersForKey(metadataEntry.getKey());
+    // get existing metadata
+    Metadata existingMetadata = getMetadata(metadataEntry.getMetadataEntity());
+
+    // the entity does not have any metadata associated with it so we should index it with its type too
+    if (existingMetadata.getProperties().isEmpty() && existingMetadata.getTags().isEmpty()) {
+      indexersForKey.add(new MetadataEntityTypeIndexer());
+    }
+    Metadata updatedMetadata = writeWithHistory(existingMetadata, metadataEntry, indexersForKey);
+    return new MetadataChange(existingMetadata, updatedMetadata);
+  }
+
+  /**
+   * Writes the entry for the {@link MetadataEntity}
+   * @param existing the existing metadata to which the metadata is being written
+   * @param entry the new metadata entry
+   * @param indexers the indexers to use for this entry
+   * @return {@link} the updated {@link Metadata}
+   */
+  private Metadata writeWithHistory(Metadata existing, MetadataEntry entry, Set<Indexer> indexers) {
+    Metadata updatedMetadata;
+    // Metadata consists of properties and tags. Properties are normal key-value pair and tag is a set of comma
+    // separated value whose key is fixes to TAGS_KEY
+    if (TAGS_KEY.equals(entry.getKey())) {
+      // we need to handle tag updates a little differently than normal key-value pair as addition of non existing
+      // tag should not overwrite all the other existing tag and also we only want to generate indexes for the
+      // newly added tags
+      Set<String> updatedTags = new HashSet<>(existing.getTags());
+      updatedTags.addAll(splitTags(entry.getValue()));
+
+      // the new metadata entry with update tags
+      MetadataEntry combined = new MetadataEntry(entry.getMetadataEntity(), entry.getKey(),
+                                                 Joiner.on(TAGS_SEPARATOR).join(updatedTags));
+      // write the value
+      writeValue(combined, indexers);
```

Comment:
these lines are very similar here and in the else block. I think it's better to move them outside the if-else. So these blocks just update the tags or the properties, and everything else is outside.

Reply:
Done.